presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStatusSender.java

```diff
@@ -0,0 +1,142 @@
+
+    private final ConcurrentMap<QueryId, PeriodicTaskExecutor> queries = new ConcurrentHashMap<>();
+
+    private PeriodicTaskExecutor nodeHeartbeatSender;
+
+    @Inject
+    public ResourceManagerClusterStatusSender(
+            @ForResourceManager DriftClient<ResourceManagerClient> resourceManagerClient,
+            InternalNodeManager internalNodeManager,
+            StatusResource statusResource,
+            @ForResourceManager ScheduledExecutorService executor,
+            ResourceManagerConfig resourceManagerConfig)
+    {
+        this.resourceManagerClient = requireNonNull(resourceManagerClient, "resourceManagerService is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        requireNonNull(statusResource, "statusResource is null");
+        this.statusSupplier = statusResource::getStatus;
+        this.executor = requireNonNull(executor, "executor is null");
+        this.resourceManagerConfig = requireNonNull(resourceManagerConfig, "resourceManagerConfig is null");
+    }
+
+    public ResourceManagerClusterStatusSender(
+            DriftClient<ResourceManagerClient> resourceManagerClient,
+            InternalNodeManager internalNodeManager,
+            Supplier<NodeStatus> statusResource,
+            ScheduledExecutorService executor,
+            ResourceManagerConfig resourceManagerConfig)
+    {
+        this.resourceManagerClient = requireNonNull(resourceManagerClient, "resourceManagerService is null");
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        this.statusSupplier = requireNonNull(statusResource, "statusResource is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        this.resourceManagerConfig = requireNonNull(resourceManagerConfig, "resourceManagerConfig is null");
+    }
+
+    @PostConstruct
+    public void init()
+    {
+        nodeHeartbeatSender = new PeriodicTaskExecutor(resourceManagerConfig.getNodeHeartbeatInterval().toMillis(), executor, this::sendNodeHeartbeat);
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        queries.values().forEach(PeriodicTaskExecutor::stop);
+        if (nodeHeartbeatSender != null) {
+            nodeHeartbeatSender.stop();
+        }
+    }
+
+    @Override
+    public void registerQuery(ManagedQueryExecution queryExecution)
+    {
+        QueryId queryId = queryExecution.getBasicQueryInfo().getQueryId();
+        queries.computeIfAbsent(queryId, unused -> new PeriodicTaskExecutor(
+                resourceManagerConfig.getQueryHeartbeatInterval().toMillis(),
+                executor,
+                () -> sendQueryHeartbeat(queryExecution)));
+        queryExecution.addStateChangeListener(newState -> {
+            if (newState.isDone()) {
+                queries.computeIfPresent(queryId, (unused, queryHeartbeatSender) -> {
+                    queryHeartbeatSender.forceRun();
```

Comment:
This is a blocking query on query completion. Will this cause query completion time to be too long (~10ms) causing long finishing time?

Reply:
Internally it's offloaded to an executor.